### PR TITLE
Fix for #550 and related

### PR DESCRIPTION
**Pull request by HuanhuanSunMSFT**
IntelliSense should not be triggered in strings or comments with only one exception if it's triggered for path in strings.
@jinglou- @AndreSayreMSFT

### CODE REVIEW

**Comment by ASayre** [comment]
Fixes #550

File: PowerShellTools/Intellisense/Utilities.cs [hunk 1]
```diff
@@ -14,52 +14,52 @@ namespace PowerShellTools.Intellisense
 {
     internal static class Utilities
     {
-        internal static bool IsGroupStart(char ch)
+        public static bool IsGroupStart(char ch)
```
Comment (ASayre) [inline PowerShellTools/Intellisense/Utilities.cs:17]: Why were these changed to Public?

File: PowerShellTools/Intellisense/IntelliSenseManager.cs [hunk 2]
```diff
@@ -641,6 +654,15 @@ private void CompletionSession_Dismissed(object sender, EventArgs e)
             _activeSession = null;
         }
 
+        private bool? IsInStringArea(bool? isInStringArea)
```
Comment (ASayre) [inline PowerShellTools/Intellisense/IntelliSenseManager.cs:657]: ?  IsInStringArea( isInStringArea )

File: PowerShellTools/Intellisense/IntelliSenseManager.cs [hunk 3]
```diff
@@ -314,7 +323,11 @@ public int Exec(ref Guid pguidCmdGroup, uint nCmdId, uint nCmdexecopt, IntPtr pv
                 (justCommitIntelliSense || (IsIntelliSenseTriggerDot(typedChar) && IsPreviousTokenVariable())) || // If dot just commit a session or previous token before a dot was a variable, trigger intellisense
                 (char.IsWhiteSpace(typedChar) && IsPreviousTokenParameter())) // If the previous token before a space was a parameter, trigger intellisense
             {
-                TriggerCompletion();
+                isInStringArea = this.IsInStringArea(isInStringArea);
+                if (isInStringArea == false)
```
Comment (ASayre) [inline PowerShellTools/Intellisense/IntelliSenseManager.cs:327]: !foo

File: PowerShellTools/Intellisense/IntelliSenseManager.cs [hunk 4]
```diff
@@ -211,7 +215,12 @@ public int Exec(ref Guid pguidCmdGroup, uint nCmdId, uint nCmdexecopt, IntPtr pv
                     break;
 
                 case VSConstants.VSStd2KCmdID.COMPLETEWORD:
-
+                    isInStringArea = this.IsInStringArea(isInStringArea);
+                    if (isInStringArea == true)
```
Comment (ASayre) [inline PowerShellTools/Intellisense/IntelliSenseManager.cs:219]: if( foo )

File: PowerShellTools/Intellisense/IntelliSenseManager.cs [hunk 5]
```diff
@@ -130,10 +131,13 @@ public int Exec(ref Guid pguidCmdGroup, uint nCmdId, uint nCmdexecopt, IntPtr pv
                 Log.DebugFormat("Typed Character: '{0}'", (typedChar == char.MinValue) ? "<null>" : typedChar.ToString());
 
                 if (_activeSession == null &&
-                    IsNotIntelliSenseTriggerWhenInStringLiteral(typedChar) &&
-                    Utilities.IsInStringArea(_textView.Caret.Position.BufferPosition.Position, _textView.TextBuffer))
+                    IsNotIntelliSenseTriggerWhenInStringLiteral(typedChar))
                 {
-                    return NextCommandHandler.Exec(ref pguidCmdGroup, nCmdId, nCmdexecopt, pvaIn, pvaOut);
+                    isInStringArea = this.IsInStringArea(isInStringArea);
+                    if (isInStringArea == true)
```
Comment (ASayre) [inline PowerShellTools/Intellisense/IntelliSenseManager.cs:137]: if( foo )

File: PowerShellTools/Intellisense/Utilities.cs [hunk 6]
```diff
@@ -103,6 +103,21 @@ internal static bool IsCaretInCommentArea(ITextView textView)
         }
 
         /// <summary>
+        /// Determins if caretis in string area.
```
Comment (ASayre) [inline PowerShellTools/Intellisense/Utilities.cs]: spelling +  String area being string literal or something else?

File: PowerShellTools/Intellisense/Utilities.cs [hunk 7]
```diff
@@ -103,6 +103,21 @@ internal static bool IsCaretInCommentArea(ITextView textView)
         }
 
         /// <summary>
+        /// Determins if caretis in string area.
+        /// </summary>
+        /// <returns></returns>
+        public static bool IsInStringArea(ITextView textView)
```
Comment (ASayre) [inline PowerShellTools/Intellisense/Utilities.cs:109]: More Q's about Public

File: PowerShellTools/Intellisense/Utilities.cs (hunk 1 above)
Comment (HuanhuanSunMSFT) [inline PowerShellTools/Intellisense/Utilities.cs:17]: I should have set them as Public from the first. Whether a method in a internal static class should be internal or public depends on when we change the class to public, do we still want the class to be internal. And that answer to be is no.

File: PowerShellTools/Intellisense/IntelliSenseManager.cs (hunk 2 above)
Comment (HuanhuanSunMSFT) [inline PowerShellTools/Intellisense/IntelliSenseManager.cs:657]: The intention is to avoid multiple invoking of the utility method to determine whether it's in string area.

File: PowerShellTools/Intellisense/IntelliSenseManager.cs (hunk 3 above)
Comment (HuanhuanSunMSFT) [inline PowerShellTools/Intellisense/IntelliSenseManager.cs:327]: This is a _bool?_

File: PowerShellTools/Intellisense/IntelliSenseManager.cs (hunk 4 above)
Comment (HuanhuanSunMSFT) [inline PowerShellTools/Intellisense/IntelliSenseManager.cs:219]: This is a bool? This way we'll have a compile error. Ref: https://msdn.microsoft.com/en-us/library/bb384091.aspx

File: PowerShellTools/Intellisense/Utilities.cs (hunk 6 above)
Comment (HuanhuanSunMSFT) [inline PowerShellTools/Intellisense/Utilities.cs]: There are two kinds of way to measure types of tokens in PowerShell, the lower level is Token which has four kinds of string type: StringLiteral = 12, StringExpandable = 13,    HereStringLiteral = 14, HereStringExpandable = 15. And another is more abstract which is PSToken, it only has one kind of string type: String. So I use string area instead of string literal here to include a more general string type.

Comment (HuanhuanSunMSFT) [inline PowerShellTools/Intellisense/Utilities.cs]: I'll just change it to: Determines if caret is in a string. This way it at least align with PSToken

**Comment by HuanhuanSunMSFT** [comment]
Also fixes #537

**Comment by ASayre** [comment]
Did this already get merged in accidentally as part of a different pull request?

**Comment by HuanhuanSunMSFT** [comment]
Looks like so.
